### PR TITLE
Port library to TypeScript

### PR DESCRIPTION
**Pull request by Favna**
I asked whether this would be appreciated on PR #40 and @sr229 said that I'd be welcome to, so here it is. First things first, some things that should be done by someone from the team with the access to TravisCI and CodeClimate for this repository:

1. Once merged please make a release tag `2.0.0`. I've added a `CHANGELOG.md` to [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and I've referenced the changes made as 2.0.0.

2. I Setup tests to only load from environment variables, to prevent the failing PR I got on #40 due to not having a `token.json` on the pipeline. This means you need to set a SauceNAO token in the travis settings, like seen below. For local running you can add it many ways, but the recommended way is documented in `tests/README.md`

1. You guys have a `.codeclimate.yml` file but the code for uploading a test coverage report to codeclimate was missing. I've added support for it from TravisCI. For this you need to add another enviroment variable to the Travis configuration. You can find the `CC_TEST_REPORTER_ID` on your codeclimate configuration panel (likely https://codeclimate.com/github/ClarityCafe/Sagiri since for me it's that but then Favna) and then going to `Repo Settings` -> `Test coverage` -> copy `TEST REPORTER ID`.
     - Sidenote that the codeclimate file was like 2 years outdated. Late 2017 codeclimate introduced a completely new config style and the file was still using the old style thus being completely useless. It's upgraded now.

image

That about covers the most important parts so lets get into the unordered list:

- Publishing of the lib is as easy as running `yarn publish` / `npm publish`. the `prepublishOnly` script automatically generates the latest build and docs
- TypeScript, obviously
  - Much improvement typings in the dist (i.e. constants actually being exported)
  - Better exports (see below)
- Included a CHANGELOG.md
- Refactored the code pretty much entirely
  - leveraging `node-fetch` rather than build in http/https libs.
  - Wrote everything using async/await
  - Wrote everything as TypeScript boosted ES6 classes
  - Both Named and Default exported Sagiri so all common methods are possible
    - Possible ways to import:

    ```ts
    // With CommonJS 
    const { Sagiri } = require('sagiri');
    const Sagiri = require('sagiri');

    // With ES Modules
    import Sagiri from 'sagiri';
    import { Sagiri } from 'sagiri';
    ```
  - Far more thorough error checking in part thanks to writing in TS (I did find various bugs as a matter of fact)
  - Ratings are now strings (like 'UNKNOWN' and 'SAFE' rather than numbers, so people actually know what the rating is without having to map them
  - Included author_url and author_name in the result set
  - Redid all the JSDoc to be compliant to jsdoc-to-markdown
- Automatic API doc generation using (jsdoc-to-markdown)[https://github.com/jsdoc2md/jsdoc-to-markdown]
  - It gets put inside the README.md through the template `docs/template.hbs`. You should not manually edit README.md as it will be reverted once the generator (`npm run docs`) is ran. Instead edit the `template.hbs` file.
- Changed ESLint config to [my own](https://github.com/favware/eslint-config/blob/master/src/index.ts) which leverages the [@typescript-eslint](https://github.com/typescript-eslint/typescript-eslint) parser, TypeScript devs have voiced that they will be using this parser later this year as well and TSLint devs have voiced they are deprectating TSLint in favour of it, so enough reason to jump to it now.
- Jest test framework for testing. Code is still mostly just the code you guys had but now put into a proper test framework.
- Rollup for bundling. Rollup is a well.. bundling library. Used by major vendors such as React and many others. It produces very minimal bundles ideal for distribution
- Removed the ternJS file. Unless one of your devs is "one of those" emacs / vim / sublime users there really is no reason to. Atom, VSCode and any other IDE have proper build-in IntelliSense ¯\\\_(ツ)_/¯ 

I think that about covers it but it's just past 5 AM here and I've been working on this since like 7 PM yesterday doing it all in 1 sitting so excuse me if I forgot something lol. If you have any questions please do ask. I've also joined the `Mya-nee Café` Discord server (Favna#0001) so you can also DM / Ping me there.

Oh and just for the lulz, this is the `dist` folder that ends up getting produced:

image

And this is (part of) `index.js`:

image

### CODE REVIEW

**Comment by Favna** [comment]
Sidenote, Travis should be failing for now but once someone performs action point 2 from the top of the post I can push a dummy change just to trigger it again and it should pass then.

File: LICENSE.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+The MIT License (MIT)
```
Comment (sr229) [inline LICENSE.md]: We really didn't need this to be styled.

File: README.md [hunk 2]
```diff
@@ -46,12 +104,14 @@ The meaning of these values are:
 
 If `options.getRatings` is not true, then this value will always be `0`.
 
+* * *
+
 ## Contributing
 
 All contributions are accepted! If you think you can bring uploading support, or make the lib perform better, make a PR and start coding!
```
Comment (sr229) [inline README.md]: You supplied a contribution guideline yet it wasn't hotlinked in the README

**Comment by sr229** [comment]
Looks like one test is failing as well, until the errors are about missing API keys I won't merge this at the moment.

**Comment by sr229** [comment]
My only issue is usually people would like to use the library out of the box without worrying about peer deps. This might've broke with our deisgn philosophy for it that it should be minimal and worry free as possible (inclusion of `node-fetch` and putting it as a peer-dep and moving `form-data` as a peer dep).

Making a critical dependency a peer dependency is kinda bad because starting from NPMv3, peer dependencies are not automatically installed.

File: README.md (hunk 2 above)
Comment (Favna) [reply README.md]: Linked it in the README as of 744f702235e2327ca141f8f4aab1ad45c5d5ab71

File: LICENSE.md (hunk 1 above)
Comment (Favna) [reply LICENSE.md]: Restored in 4d6f07a8cdee5a208770942c606f96a100f51402

**Comment by Favna** [comment]
A good sleep later I managed to have a clear head and get the dependencies (node-fetch and form-data) as actual dependencies instead of peerdependencies so the end-user won't have to install them manually anymore.

**Comment by Favna** [comment]
I see I forgot to mention that I set Travis to Node 10 and Node 12 only. Can't really do more than 2 as there are 3 tests which just exactly fits in the 6 calls per 30 seconds ratelimit of SauceNAO for free accounts, so sadly if more tests need to be configured then there has to be a token in Travis of an account with more than just SauceNao basic. The reason I went for 10 and 12 is because those are the current LTS and Latest respectively.

**Comment by sr229** [comment]
> I see I forgot to mention that I set Travis to Node 10 and Node 12 only. Can't really do more than 2 as there are 3 tests which just exactly fits in the 6 calls per 30 seconds ratelimit of SauceNAO for free accounts, so sadly if more tests need to be configured then there has to be a token in Travis of an account with more than just SauceNao basic. The reason I went for 10 and 12 is because those are the current LTS and Latest respectively.

Good call, since I think there's no more issues I'll gladly merge it later.

**Pull request merged**